crates/iota-data-ingestion-core/src/reducer.rs

```diff
@@ -144,6 +151,79 @@ pub(crate) async fn reduce<W: Worker>(
                 .map_err(|_| IngestionError::Channel("unable to send worker pool progress updates to executor, receiver half closed".into()))?;
             progress_update = None;
         }
+
+        // Check for shutdown signal after progress update to ensure progress
+        // is reported even during shutdown.
+        if trigger_shutdown {
+            break;
+        }
     }
     Ok(())
 }
+
+/// Represents the outcome of a commit operation with retry logic.
+enum CommitStatus {
+    /// Commit succeeded, continue processing.
+    Success,
+    /// Processing should stop due to shutdown signal.
+    Shutdown,
+}
+
+/// Attempts to commit a batch of messages with exponential backoff retries on
+/// failure.
+///
+/// This function repeatedly calls the [`commit`](Reducer::commit) method of the
+/// provided [`Reducer`] until either:
+/// - The commit succeeds, returning `CommitStatus::Success`.
+/// - A cancellation signal is received via the [`CancellationToken`], returning
+///   `CommitStatus::Shutdown`.
+/// - All retry attempts are exhausted within a 15-minute maximum elapsed time,
+///   causing a panic.
+///
+/// # Retry Mechanism:
+/// - Uses [`ExponentialBackoff`](backoff::ExponentialBackoff) to introduce
+///   increasing delays between retry attempts.
+/// - Checks for cancellation both before and after each commit attempt.
+/// - If a cancellation signal is received during a backoff delay, the function
+///   exits immediately with `CommitStatus::Shutdown`.
+///
+/// # Panics:
+/// - If all retry attempts are exhausted within the 15-minute maximum elapsed
+///   time, indicating a persistent failure.
+async fn commit_with_retry<W: Worker>(
+    reducer: &dyn Reducer<W>,
+    batch: Arc<Vec<<W as Worker>::Message>>,
```

Comment:
Here on the other hand you can consume so that `batch: Vec<T>`, no?

Reply:
fixed: https://github.com/iotaledger/iota/commit/a36cb9392bd604fbd8669ecf6d81fa044ef1257c